src/ray/core_worker/experimental_channel_manager.h

```diff
@@ -0,0 +1,146 @@
+  /// header, which is used to synchronize with other writers and readers, and
+  /// the object data and metadata, which is read by the application.
+  /// \return The return status.
+  Status RegisterWriterChannel(const ObjectID &object_id,
+                               std::unique_ptr<plasma::MutableObject> mutable_object);
+
+  /// Acquires a write lock on the object that prevents readers from reading
+  /// until we are done writing. This is safe for concurrent writers.
+  ///
+  /// \param[in] object_id The ID of the object.
+  /// \param[in] data_size The size of the object to write. This overwrites the
+  /// current data size.
+  /// \param[in] metadata A pointer to the object metadata buffer to copy. This
+  /// will overwrite the current metadata.
+  /// \param[in] metadata_size The number of bytes to copy from the metadata
+  /// pointer.
+  /// \param[in] num_readers The number of readers that must read and release
+  /// \param[out] data The mutable object buffer in plasma that can be written to.
+  /// \return The return status.
+  Status WriteAcquire(const ObjectID &object_id,
+                      int64_t data_size,
+                      const uint8_t *metadata,
+                      int64_t metadata_size,
+                      int64_t num_readers,
+                      std::shared_ptr<Buffer> *data);
+
+  /// Releases an acquired write lock on the object, allowing readers to read.
+  /// This is the equivalent of "Seal" for normal objects.
+  ///
+  /// \param[in] object_id The ID of the object.
+  /// \return The return status.
+  Status WriteRelease(const ObjectID &object_id);
+
+  /// Register the caller as a writer for the channel.
+  ///
+  /// \param[in] object_id The ID of the object.
+  /// \param[in] mutable_object Struct containing pointers for the object
+  /// header, which is used to synchronize with other writers and readers, and
+  /// the object data and metadata, read by the application.
+  /// \return The return status.
+  Status RegisterReaderChannel(const ObjectID &object_id,
+                               std::unique_ptr<plasma::MutableObject> mutable_object);
+
+  /// Return whether the caller is registered as a reader for this object.
+  ///
+  /// \return Whether the reader is registered.
+  bool ReaderChannelRegistered(const ObjectID &object_id) const;
+
+  /// Acquires a read lock on the object that prevents the writer from writing
+  /// again until we are done reading the current value.
+  ///
+  /// \param[in] object_id The ID of the object.
+  /// \param[out] result The read object. This buffer is guaranteed to be valid
+  /// until the caller calls ReadRelease next.
+  /// \return The return status. The ReadAcquire can fail if there have already
+  /// been `num_readers` for the current value.
+  Status ReadAcquire(const ObjectID &object_id, std::shared_ptr<RayObject> *result);
+
+  /// Releases the object, allowing it to be written again. If the caller did
```

Comment:
> If the caller did
  /// not previously ReadAcquire the object, then this first blocks until the
  /// latest value is available to read, then releases the value.

What's the use case of this behavior?

Reply:
Mostly just for convenience / staying in C++, so we can release the values without having to explicitly read them in python.